### PR TITLE
Feature: store learns to delete prefixes when overwriting/creating hierarchy nodes

### PR DESCRIPTION
**Pull request by jhamman**
This PR implements `delete_dir` teaches the Group and Array classes to use it when overwriting nodes in a hierarchy. 

Also included here:
~- `delete_prefix` - not used but perhaps we should just use it instead of `delete_dir`~
- changes `list_prefix` to return absolute paths instead of stripping `prefix`. Given that `list_prefix` was unused before, this wasn't very disruptive but calling it out so others can help decide if this is what we want.

Still needs:
- [x] new store tests for `delete_dir`
- [x] new group tests for edge cases around `delete_dir`

closes https://github.com/zarr-developers/zarr-python/issues/2191
first steps toward https://github.com/zarr-developers/zarr-python/issues/2108 and #2359 

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

File: src/zarr/abc/store.py [hunk 1]
```diff
@@ -371,6 +371,36 @@ def list_dir(self, prefix: str) -> AsyncGenerator[str, None]:
...
+        if not self.supports_deletes:
+            raise NotImplementedError
+        if not self.supports_listing:
+            raise NotImplementedError
+        self._check_writable()
+        if recursive:
+            if not prefix.endswith("/"):
+                prefix += "/"
+            async for key in self.list_prefix(prefix):
+                await self.delete(f"{key}")
```
Comment (d-v-b) [inline src/zarr/abc/store.py:387]: do we need an fstring here?

File: src/zarr/api/asynchronous.py [hunk 2]
```diff
@@ -712,7 +712,7 @@ async def create(
     dtype: npt.DTypeLike | None = None,
     compressor: dict[str, JSON] | None = None,  # TODO: default and type change
     fill_value: Any | None = 0,  # TODO: need type
-    order: MemoryOrder | None = None,  # TODO: default change
+    order: MemoryOrder | None = None,
```
Comment (d-v-b) [inline src/zarr/api/asynchronous.py:715]: weird that these changes from #2417 are appearing here.

File: src/zarr/core/array.py [hunk 3]
```diff
@@ -554,7 +554,12 @@ async def _create_v3(
         attributes: dict[str, JSON] | None = None,
         exists_ok: bool = False,
     ) -> AsyncArray[ArrayV3Metadata]:
-        if not exists_ok:
+        if exists_ok:
+            if store_path.store.supports_deletes:
+                await store_path.delete_dir(recursive=True)
+            else:
+                await ensure_no_existing_node(store_path, zarr_format=3)
+        else:
```
Comment (d-v-b) [inline src/zarr/core/array.py:562]: [not a blocker for this PR but just an idea triggered by this code]

deleting an entire array can take `O(num_chunks)` time, and I know this has been a pain point for for zarr users including myself. To alleviate this, we should expose two settings:
- one setting for array creation that avoids a recursive delete, and instead only creates / modifies metadata documents
- one setting for chunks that causes zarr to ignore existing invalid chunks and simply overwrite them.

together, these two settings would make creating re-creating arrays in-place very cheap, albeit more likely to contain garbage from old arrays / groups, so if we do this we should also create a routine for cleaning up that garbage.

File: src/zarr/storage/local.py [hunk 4]
```diff
@@ -226,7 +226,9 @@ async def list(self) -> AsyncGenerator[str, None]:
 
     async def list_prefix(self, prefix: str) -> AsyncGenerator[str, None]:
         # docstring inherited
-        to_strip = os.path.join(str(self.root / prefix))
+        to_strip = str(self.root)
+        if prefix.endswith("/"):
+            prefix = prefix[:-1]
```
Comment (d-v-b) [inline src/zarr/storage/local.py:231]:
```suggestion
            prefix = prefix.rstrip('/')
```

File: src/zarr/core/array.py (hunk 3 above)
Comment (jhamman) [reply src/zarr/core/array.py:562]: Happy to discuss this idea in a separate issue by my initial reaction is that this sounds like a very dangerous option!

> To alleviate this, we should expose two settings:

The big change in 3.0 is that the store is async. So we can optimize this by more aggressively utilizing concurrency.

Comment (d-v-b) [reply src/zarr/core/array.py:562]: > Happy to discuss this idea in a separate issue by my initial reaction is that this sounds like a very dangerous option!

I will open an issue for the idea. I agree that it's opens up a "hard hat zone" in the API. 👷

File: src/zarr/abc/store.py [hunk 5]
```diff
@@ -371,6 +371,36 @@ def list_dir(self, prefix: str) -> AsyncGenerator[str, None]:
...
+        if recursive:
+            if not prefix.endswith("/"):
+                prefix += "/"
+            async for key in self.list_prefix(prefix):
+                await self.delete(key)
+        else:
+            async for key in self.list_dir(prefix):
+                await self.delete(f"{prefix}/{key}")
+
+    async def delete_prefix(self, prefix: str) -> None:
```
Comment (jhamman) [inline src/zarr/abc/store.py:392]: Open to input on this one but I do not think this is needed and could be removed.

File: src/zarr/abc/store.py [hunk 6]
```diff
@@ -371,6 +371,36 @@ def list_dir(self, prefix: str) -> AsyncGenerator[str, None]:
         """
         ...
 
+    async def delete_dir(self, prefix: str, recursive: bool = True) -> None:
+        """
+        Remove all keys and prefixes in the store that begin with a given prefix.
```
Comment (paraseba) [inline src/zarr/abc/store.py:376]: I think it would make life much easier for `Store` implementers if we declare that `prefix` will always be the path of a group or a array. Otherwise, all implementations need to deal with unnecessary edge cases like `delete_dir("path/to/array/c/0")` or worse.

[Icechunk](https://icechunk.io/) is an example of a store that could very cleanly and efficiently implement the case for a group/array prefix, but it would struggle to handle other prefixes.

File: src/zarr/abc/store.py (hunk 5 above)
Comment (paraseba) [reply src/zarr/abc/store.py:392]: I'd delete if we don't need it.

File: src/zarr/abc/store.py (hunk 6 above)
Comment (paraseba) [reply src/zarr/abc/store.py:376]: What is the use case for `recursive = False` ? If zarr itself doesn't have one, I'd remove the argument.

File: src/zarr/abc/store.py [hunk 7]
```diff
@@ -371,6 +371,36 @@ def list_dir(self, prefix: str) -> AsyncGenerator[str, None]:
...
+    async def delete_prefix(self, prefix: str) -> None:
+        """
+        Remove all keys in the store that begin with a given prefix.
+        """
+        if not self.supports_deletes:
+            raise NotImplementedError
+        if not self.supports_listing:
+            raise NotImplementedError
+        self._check_writable()
+        async for key in self.list_prefix(prefix):
```
Comment (paraseba) [inline src/zarr/abc/store.py:401]: Can we document what's the expected return value of `list_prefix`? Absolute vs. relative, final "/" or not, etc.

File: src/zarr/abc/store.py (hunk 6 above)
Comment (jhamman) [reply src/zarr/abc/store.py:376]: I can see a use case for `delete_dir("path/to/array/c/")` (remove all chunks from an array) so I think we should allow that -- even if it makes life a little harder in icechunk.

File: src/zarr/abc/store.py (hunk 7 above)
Comment (jhamman) [reply src/zarr/abc/store.py:385]: this is done in the list_prefix docstring

**Comment by TomAugspurger** [comment]
Looks good overall. I've been struggling a bit with whether Stores should be a bit more aware about the context the operation is occurring in (e.g. deleting a group vs a specific chunk of an array; I think this relates to the conversation at https://github.com/zarr-developers/zarr-python/pull/2430#discussion_r1813941831). But I think this is a good improvement on its own.

File: src/zarr/abc/store.py (hunk 6 above)
Comment (paraseba) [reply src/zarr/abc/store.py:376]: That's totally fine. Let's document the three possible prefix cases, and say that stores are allowed to do undefined behavior for other prefixes

**Review by d-v-b — APPROVED** [review]
looks good. regarding the change to `list_prefix`, I will note that the definition of "absolute paths" is a bit shaky across different stores, since `RemoteStore` will return keys relative to its `path` attribute, and similarly `LocalStore` with its `root` attribute.

**Comment by jhamman** [comment]
> regarding the change to list_prefix, I will note that the definition of "absolute paths" is a bit shaky across different stores

Would defining these as "relative to the root of the store" be clearer?

File: src/zarr/abc/store.py [hunk 8]
```diff
@@ -342,8 +342,8 @@ def list(self) -> AsyncGenerator[str, None]:
     @abstractmethod
     def list_prefix(self, prefix: str) -> AsyncGenerator[str, None]:
         """
-        Retrieve all keys in the store that begin with a given prefix. Keys are returned with the
-        common leading prefix removed.
+        Retrieve all keys in the store that begin with a given prefix. Keys are returned as
+        absolute paths (i.e. including the prefix).
```
Comment (jhamman) [inline src/zarr/abc/store.py:346]:
```suggestion
        Retrieve all keys in the store that begin with a given prefix. Keys are returned relative 
        to the root of the store.
```

**Comment by jhamman** [comment]
pre-commit.ci autofix
